[Merged by Bors] - Enable deriving Reflect on structs with generic types

cBournhonesque opened the pull request:
# Objective

I recently had an issue, where I have a struct:
```
struct Property {
   inner: T
}
```
that I use as a wrapper for internal purposes.
I don't want to update my struct definition to 
```
struct Property<T: Reflect>{
   inner: T
}
```
because I still want to be able to build `Property<T>` for types `T` that are not `Reflect`. (and also because I don't want to update my whole code base with `<T: Reflect>` bounds)

I still wanted to have reflection on it (for `bevy_inspector_egui`), but adding `derive(Reflect)` fails with the error:
`T cannot be sent between threads safely. T needs to implement Sync.`

I believe that `bevy_reflect` should adopt the model of other derives in the case of generics, which is to add the `Reflect` implementation only if the generics also implement `Reflect`. (That is the behaviour of other macros such as `derive(Clone)` or `derive(Debug)`.

It's also the current behavior of `derive(FromReflect)`.

Basically doing something like:
```
impl<T> Reflect for Foo<T>
where T: Reflect
```


## Solution

- I updated the derive macros for `Structs` and `TupleStructs` to add extra `where` bounds.
   -  Every type that is reflected will need a `T: Reflect` bound
   -  Ignored types will need a `T: 'static + Send + Sync` bound. Here's the reason. For cases like this:
```
#[derive(Reflect)]
struct Foo<T, U>{
   a: T
   #[reflect(ignore)]
   b: U
}
```
I had to add the bound `'static + Send + Sync` to ignored generics like `U`.
The reason is that we want `Foo<T, U>` to be `Reflect: 'static + Send + Sync`, so `Foo<T, U>` must be able to implement those auto-traits. `Foo<T, U>` will only implement those auto-traits if every generic type implements them, including ignored types.
This means that the previously compile-fail case now compiles:
```
#[derive(Reflect)]
struct Foo<'a> {
    #[reflect(ignore)]
    value: &'a str,
}
```
But `Foo<'a>` will only be useable in the cases where `'a: 'static` and panic if we don't have `'a: 'static`, which is what we want (nice bonus from this PR ;) )



---

## Changelog

> This section is optional. If this was a trivial fix, or has no externally-visible impact, you can delete this section.

### Added
Possibility to add `derive(Reflect)` to structs and enums that contain generic types, like so:
```
#[derive(Reflect)]
struct Foo<T>{
   a: T
}
```
Reflection will only be available if the generic type T also implements `Reflect`.
(previously, this would just return a compiler error)

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/registration.rs:
```diff
@@ -3,14 +3,15 @@
     bevy_reflect_path: &Path,
     registration_data: &[Ident],
     generics: &Generics,
+    field_types: &Vec<Type>,
```
Nit: I don't think we need to add this. We should be able to get the generic param info from `generics`, right?

MrGVSV replied:
Oh I think I see why we do it this way. This ensures we only add the bound to actively used (non-`#[reflect(ignore)]`) types.

cBournhonesque replied:
Yes exactly; we want to ignore generics that are inactive

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/registration.rs:
```diff
@@ -22,9 +23,21 @@ pub(crate) fn impl_get_type_registration(
     });
 
+    // Add Typed bound for each active field
+    let mut where_reflect_typed_clause = if where_clause.is_some() {
+        quote! {#where_clause}
+    } else if !field_types.is_empty() {
+        quote! {where}
+    } else {
+        quote! {}
+    };
+    where_reflect_typed_clause.extend(quote! {
+        #(#field_types: #bevy_reflect_path::Reflect + #bevy_reflect_path::Typed,)*
+    });
```
Since this is used throughout multiple modules now, could we add it to `bevy_reflect_derive/src/utility.rs`?

Alternatively, we could add an even more general-purpose version of this to `bevy_macro_utils`.

cBournhonesque replied:
Good idea!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/impls/structs.rs:
```diff
@@ -99,16 +100,34 @@ pub(crate) fn impl_struct(reflect_struct: &ReflectStruct) -> TokenStream {
+        quote! {}
+    };
+    where_reflect_clause.extend(quote! {
+        // TODO: had to add #bevy_reflect_path::Typed to get the test to compile,
+        //  presumably because of this:
+        //  #[inline]
+        //  fn get_type_info(&self) -> &'static bevy_reflect::TypeInfo {
+        //      <Self as bevy_reflect::Typed>::type_info()
+        //  }
```
Is there no other way around this? We shouldn't need to specify this since the `Typed` impl guarantees `Self: Typed`. Or maybe I'm missing something 🤔

cBournhonesque replied:
I think it's because in the `Typed` derive, I had to add a `where T: Typed` clause.
It was not clear to me why I needed it.
Something in here:
```
            fn type_info() -> &'static bevy_reflect::TypeInfo {
                static CELL: bevy_reflect::utility::NonGenericTypeInfoCell = bevy_reflect::utility::NonGenericTypeInfoCell::new();
                CELL.get_or_set(|| {
                    let fields = [
                        bevy_reflect::NamedField::new::<u32>("a"),
                        bevy_reflect::NamedField::new::<String>("b"),
                    ];
                    let info = bevy_reflect::StructInfo::new::<Self>("Foo", &fields);
                    bevy_reflect::TypeInfo::Struct(info)
                })
            }
```
required it.

cBournhonesque replied:
For example, if I try with
```
#[derive(Reflect)]
struct Foo<T, U> {
    a: T,
    #[reflect(ignore)]
    b: U
}
```
 I get this error still:
```
error[E0277]: `U` cannot be shared between threads safely
164 |         struct Foo<T, U> {
    |                ^^^
note: required by a bound in `type_info::Typed`
   --> crates/bevy_reflect/src/type_info.rs:69:18
    |
69  | pub trait Typed: Reflect {
    |                  ^^^^^^^ required by this bound in `Typed`
```

MrGVSV commented:
> I didn't implement this for `Enums`, because enums seem to have a slightly different code structure and it would have been a bigger change. Maybe it can be added in a different PR?

It shouldn't be too difficult to add an `active_types` method on `ReflectEnum` and/or `EnumVariant`. Doing so should make the implementation of this PR work more-or-less like the other impls.

MrGVSV reviewed (changes requested):
I think now that we have `bevy_reflect_compile_fail_tests`, we should add tests to it for PRs like this to ensure that which should fail to compile continues to fail and that which should compile continues to compile.

If there's not a failure case to test, then this comment can be ignored.

cBournhonesque commented:
@MrGVSV I made some changes:
- wrote a utility function for `where_clause`
- made sure tests pass even for ignored reflected types. This has some nice unintended consequences (see explanation in `#Solution`)
- Removed the need to derive `Typed` at all
- added a compile-fail test, for cases where the generic type doesn't implement Reflect.


There are just 2 things I am not sure about:
- I wasn't able to get it working for `Enums`. I didn't really understand how to get a type from an `EnumVariant`.
- I didn't add any bound for `ReflectRef::Value` because I don't understand what that is

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:
```diff
@@ -81,6 +81,18 @@ pub(crate) struct ReflectEnum<'a> {
     variants: Vec<EnumVariant<'a>>,
 }
 
+impl<'a> ReflectEnum<'a> {
+    /// Get a collection of types which are exposed to the reflection API
+    pub fn active_types(&self) -> Vec<syn::Type> {
+        Vec::default()
+    }
+
+    /// Get a collection of types which are ignored by the reflection API
+    pub fn ignored_types(&self) -> Vec<syn::Type> {
+        Vec::default()
+    }
+}
```
I think that if we want this to work directly on `ReflectEnum`, we need it to work on `EnumVariant`. So we could do something like:

```rust
const EMPTY_FIELDS: &'static [StructField<'static>] = &[];

impl<'a> ReflectEnum<'a> {
    pub fn active_fields(&self) -> impl Iterator<Item = &StructField<'a>> {
        self.variants()
            .iter()
            .flat_map(|variant| variant.active_fields())
    }

    pub fn active_types(&self) -> Vec<syn::Type> {
        self.active_fields().map(|field| field.data.ty.clone())
            .collect()
    }
}

impl<'a> EnumVariant<'a> {
    pub fn active_fields(&self) -> impl Iterator<Item = &StructField<'a>> {
        match &self.fields {
            EnumVariantFields::Named(fields) | EnumVariantFields::Unnamed(fields) => fields.iter(),
            EnumVariantFields::Unit => EMPTY_FIELDS.iter(),
        }
    }

    pub fn active_types(&self) -> Vec<syn::Type> {
        self.active_fields()
            .map(|field| field.data.ty.clone())
            .collect()
    }
}
```

And of course, we'd need to include similar methods for the ignored fields.

Not sure if this actually works or not so we'll need to add some tests to check that.

cBournhonesque replied:
I'd prefer working on Enums in a different PR, if you don't mind?

Mainly because `#[reflect(ignore)]` isn't supported for enums right now; so i'd like to have a separate PR that both adds
`#[reflect(ignore)` for enums and enables reflection with generics.

cBournhonesque replied:
Actually, I guess we can just make it work for active fields, and ignore ignored_fields for now; and then work on ignored fields in a future PR!

MrGVSV replied:
I just checked and it looks like `#[reflect(ignore)]` does work on enum fields? Could you verify this? I think we may have gotten it wrong on discord

MrGVSV replied:
Also I was looking at #6767 and realized we actually don't need `EMPTY_FIELDS` if we just return `&[StructField]` directly instead of an iterator

cBournhonesque replied:
Really? I tested it locally and the `#[reflect(ignore)]` fields were still in the reflect impl

cBournhonesque replied:
Got enums working :)

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:
```diff
@@ -350,14 +364,20 @@ impl<'a> ReflectStruct<'a> {
-    pub fn get_type_registration(&self) -> proc_macro2::TokenStream {
+    pub fn get_type_registration(
+        &self,
+        field_types: &Vec<Type>,
+        active_types: &Vec<Type>,
```
Nit: Maybe add some documentation for these new parameters?

cBournhonesque commented on crates/bevy_reflect_compile_fail_tests/tests/reflect_derive/lifetimes.fail.rs:
```diff
@@ -1,9 +0,0 @@
-use bevy_reflect::Reflect;
```
Yes, this doesn't fail anymore; the struct implements Reflect correctly, but can only be used for cases where `'a: 'static`

MrGVSV replied:
Oh this is because the changes make it so that something like `&'a str` is now bounded by `&'a str: 'static`? Or something like that?

cBournhonesque replied:
Yes exactly!

cBournhonesque commented on crates/bevy_reflect_compile_fail_tests/tests/reflect_derive/generics.fail.rs:
```diff
@@ -0,0 +1,15 @@
+use bevy_reflect::Reflect;
```
Sounds good! Out of curiosity, what is the benefit of having `generics.pass.rs` as opposed to just writing a test in the `lib.rs` of the main crate?

Is it for cases where we don't necessarily test anything but just want to make sure that things compile?

cBournhonesque replied:
Oh, read your other comment. Makes sense to me!

cBournhonesque commented:
Sorry I realize now that my PR was a bit hasty and lacked polishing. Thank you for the excellent comments, I agree with everything!

MrGVSV commented:
> Sorry I realize now that my PR was a bit hasty and lacked polishing. Thank you for the excellent comments, I agree with everything!

No worries! It doesn't have to be "polished" haha just as long as you’re receptive to review/discussion— which you are 😄

If anything, I should be apologizing for all my nitpicks because I realize I may have gone a bit overboard in some cases haha

cBournhonesque commented on crates/bevy_reflect/bevy_reflect_derive/src/registration.rs:
```diff
@@ -22,9 +25,18 @@ pub(crate) fn impl_get_type_registration(
+        field_types,
+        quote! { #bevy_reflect_path::Reflect },
+        ignored_types,
+        quote! { 'static + std::marker::Send + std::marker::Sync },
```
Hmm, I'd rather keep using the TokenStream for simplicity? 
I switched to `#FQAny + #FQSend + #FQSync` so it should be less error-prone

cBournhonesque commented on crates/bevy_reflect_compile_fail_tests/tests/reflect_derive/generics.fail.rs:
```diff
@@ -0,0 +1,15 @@
+struct NoReflect(f32);
+
+fn main() {
+    let mut foo = Foo::<NoReflect> { a: NoReflect(42.0) };
```
Great idea! Makes the errors much more explicit and clear !

cBournhonesque commented on crates/bevy_reflect/bevy_reflect_derive/src/impls/values.rs:
```diff
@@ -21,9 +21,13 @@ pub(crate) fn impl_value(meta: &ReflectMeta) -> TokenStream {
     #[cfg(not(feature = "documentation"))]
     let with_docs: Option<proc_macro2::TokenStream> = None;
 
+    let field_types = Vec::default();
+    let ignored_types = Vec::default();
     let typed_impl = impl_typed(
         type_name,
         meta.generics(),
+        &field_types,
+        &ignored_types,
```
SG to skip adding the bounds for `values.rs`.
I can't think of a case where they would be needed anyway.

alice-i-cecile reviewed (approved):
Well-motivated, good analysis of the behavior, solid doc strings and a decent set of compile tests. LGTM!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/utility.rs:
```diff
@@ -59,6 +60,59 @@ pub(crate) fn ident_or_index(ident: Option<&Ident>, index: usize) -> Member {
 
+/// Extends the `where` clause in reflection with any additional bounds needed.
+///
+/// * `where_clause`: existing `where` clause present on the object to be derived
```
Nit: You can check elsewhere in the crate/repo to verify, but I think the parameter list is usually sectioned off under `# Parameters`.

Same goes for examples, which are usually given the `# Example`/`# Examples` header.

If you make this change, you may need to reorganize so that your detail paragraphs come before the `# Parameters` and `# Example` section.

Then again, this is all just internal documentation so style isn't of the highest priority haha

MrGVSV reviewed (changes requested):
Awesome! Just a couple super small fixes and I think we'll be good to go!

cBournhonesque commented on crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:
```diff
@@ -2,12 +2,13 @@ use crate::container_attributes::ReflectTraits;
 use crate::field_attributes::{parse_field_attrs, ReflectFieldAttr};
 use crate::utility::members_to_serialization_denylist;
 use bit_set::BitSet;
+use proc_macro2::TokenStream;
```
We can; but in general I think the common practice is to `improt proc_macro2::TokenStream` everywhere (since it's compatible with `syn` and `quote`) and only at the very last stage convert into a `proc_macro::TokenStream`.

I think most crates have this model; where `proc_macro::TokenStream` is the only one that is explicitly spelt out

cBournhonesque replied:
Basically, I think that `use proc_macro2::TokenStream` should be used everywhere by default (since it's the one we apply ops on), and then we just call `proc_macro::TokenStream` when needed. But this can be left to another PR, ot unblock this one

MrGVSV replied:
Yeah I agree, it makes more sense to flip it. But just for consistency until a cleanup PR comes along to handle it, we should try to keep it consistent

alice-i-cecile commented:
bors try

MrGVSV reviewed (approved):
Awesome job! Thanks for making all those changes. LGTM!

alice-i-cecile commented:
bors r+
